File: typescript/src/core/damage.ts
import templates from "../assets/templates.json";
import {
  type Image,
  type Region,
  Matcher,
  resize,
  sample,
  cropGray,
  grayscale,
} from "./image";
import {
  type DamageEvent,
  type Clip,
  type DamageOptions,
  round,
} from "../shared";

export const methods = ["white_180", "gray_otsu", "white_otsu"] as const;
type Method = (typeof methods)[number];
export const banks: Record<string, Uint8Array[]> = Object.fromEntries(
  methods.map((method) => [
    method,
    Array.from({ length: 10 }, (_, n) => {
      const hex =
          templates.TEMPLATES[
            `${method}_${n}` as keyof typeof templates.TEMPLATES
          ],
        bytes = Buffer.from(hex, "hex");
      return Uint8Array.from(
        { length: 600 },
        (_, i) => (bytes[i >> 3] >> (7 - (i % 8))) & 1,
      );
    }),
  ]),
);
const icons: Image[] = templates.LOCATION_ICONS.map((hex) => ({
  data: new Uint8Array(Buffer.from(hex, "hex")),
  width: 64,
  height: 38,
  channels: 1,
}));
export interface Anchor {
  x: number;
  y: number;
  scale: number;
  score: number;
}
// Anchors stay in original-video coordinates when FFmpeg supplies only a crop.
export interface HudFrame extends Image {
  origin?: { x: number; y: number };
}
// Favor recall: a plausible icon is enough to attempt reading the digits.
const ICON_MIN_SCORE = 0.60;
const resized = new Map<string, Image>();
function search(
  gray: Image,
  scales: number[],
  origin = { x: 0, y: 0 },
  peaks = 1,
): Anchor[] {
  const matcher = new Matcher(gray),
    hits: Anchor[] = [],
    seen = new Set<string>();
  for (const scale of scales) {
    const w = Math.round(64 * scale),
      h = Math.round(38 * scale),
      size = `${w}:${h}`;
    if (w < 8 || h < 5 || seen.has(size) || w > gray.width || h > gray.height)
      continue;
    seen.add(size);
    for (let k = 0; k < icons.length; k++) {
      const key = `${k}:${size}`;
      if (!resized.has(key)) resized.set(key, resize(icons[k], w, h));
      const scores = matcher.scores(resized.get(key)!, key);
      for (let p = 0; p < peaks && scores.data.length; p++) {
        let best = 0;
        for (let i = 1; i < scores.data.length; i++)
          if (scores.data[i] > scores.data[best]) best = i;
        const x = best % scores.width,
          y = Math.floor(best / scores.width);
        hits.push({
          x: x + origin.x,
          y: y + origin.y,
          scale: w / 64,
          score: scores.data[best],
        });
        for (
          let yy = Math.max(0, y - h);
          yy <= Math.min(scores.height - 1, y + h);
          yy++
        )
          scores.data.fill(
            -1,
            yy * scores.width + Math.max(0, x - w),
            yy * scores.width + Math.min(scores.width, x + w + 1),
          );
      }
    }
  }
  return hits.sort((a, b) => b.score - a.score);
}
export class HudLocator {
  anchor: Anchor | null = null;
  private lastAnchor: Anchor | null = null;
  private recalibrateScale = false;
  private nextSearch = 0;
  private missedSearches = 0;
  searches = 0;
  localRecoveries = 0;
  locatedFrames = 0;
  constructor(private fps: number) {}
  seed(anchor: Anchor) {
    this.anchor = this.lastAnchor = { ...anchor };
  }
  // Covers every local search and the digit band. An extra search radius gives
  // the supply crop room for nearby movement without changing matching bounds.
  searchRegion(width: number, height: number, withMargin = false): Region | null {
    const a = this.lastAnchor;
    if (!a) return null;
    const radius = Math.max(24, 192 * a.scale) * (withMargin ? 2 : 1),
      x = Math.max(0, Math.floor(a.x - radius)),
      y = Math.max(0, Math.floor(a.y - radius)),
      // A recovered icon may sit at the rightmost search position. Include
      // its complete digit band at the largest scale the local retry can test.
      right = Math.min(width, Math.ceil(a.x + 64 * a.scale + radius +
        120 * Math.min(2, a.scale * 1.15) + 2)),
      bottom = Math.min(height, Math.ceil(a.y + 38 * a.scale + radius));
    return { x, y, width: right - x, height: bottom - y };
  }
  // Unreadable digits can drop the active track, but the acquired area stays
  // locked for this video. Recovery must never enable another global search.
  invalidate() {
    this.anchor = null;
    this.recalibrateScale = true;
    this.nextSearch = 0;
    this.missedSearches = 0;
  }
  private accept(anchor: Anchor): Anchor {
    this.anchor = this.lastAnchor = anchor;
    this.missedSearches = 0;
    this.locatedFrames++;
    return anchor;
  }
  private refine(
    frame: HudFrame,
    anchor: Anchor,
    radius: number,
    varyScale = true,
  ): Anchor | undefined {
    const s = anchor.scale,
      ox = frame.origin?.x ?? 0,
      oy = frame.origin?.y ?? 0,
      x = Math.max(ox, Math.floor(anchor.x - radius)),
      y = Math.max(oy, Math.floor(anchor.y - radius));
    const x1 = Math.min(ox + frame.width, Math.ceil(anchor.x + 64 * s + radius)),
      y1 = Math.min(oy + frame.height, Math.ceil(anchor.y + 38 * s + radius));
    if (x1 <= x || y1 <= y) return undefined;
    const scales = (
      varyScale
        ? Array.from({ length: 13 }, (_, i) => s * (0.85 + (0.3 * i) / 12))
        : [s]
    ).filter((v) => v >= 0.25 && v <= 2);
    return search(cropGray(frame, x - ox, y - oy, x1 - x, y1 - y), scales, { x, y })[0];
  }
  locate(frame: HudFrame, index: number): Anchor | null {
    for (const hit of this.candidateAnchors(frame, index)) {
      if (this.lastAnchor || readLocatedCounter(frame, hit) !== null)
        return this.accept(hit);
    }
    return null;
  }
  // Production OCR acquisition validates candidates with the supplied reader;
  // digit templates are only retained for the legacy detector above.
  async locateAndRead(frame: HudFrame, index: number,
    read: (anchor: Anchor) => Promise<number | null>): Promise<{ anchor: Anchor; value: number | null } | null> {
    for (const hit of this.candidateAnchors(frame, index)) {
      const value = await read(hit);
      if (this.lastAnchor || value !== null)
        return { anchor: this.accept(hit), value };
    }
    return null;
  }
  private *candidateAnchors(frame: HudFrame, index: number): Generator<Anchor> {
    if (frame.origin && !this.lastAnchor)
      throw new Error("首次 HUD 定位需要完整画面");
    const tracking = this.anchor !== null;
    if (this.anchor) {
      const hit = this.refine(
        frame,
        this.anchor,
        Math.max(12, 64 * this.anchor.scale),
        false,
      );
      if (hit && hit.score >= ICON_MIN_SCORE) {
        yield hit;
        return;
      }
      this.anchor = null;
    }
    // Keep the last location through missing HUD frames. Check it at the
    // sampling rate, even while a wider local retry is deferred.
    if (!tracking && this.lastAnchor) {
      // If the icon matched but digits stayed unreadable, recheck its scale
      // locally instead of accepting the same stale scale indefinitely.
      const hit = this.refine(
        frame,
        this.lastAnchor,
        Math.max(12, 64 * this.lastAnchor.scale),
        this.recalibrateScale,
      );
      this.recalibrateScale = false;
      if (hit && hit.score >= ICON_MIN_SCORE) {
        this.localRecoveries++;
        this.nextSearch = index + Math.max(1, Math.round(this.fps));
        yield hit;
        return;
      }
    }
    if (index < this.nextSearch) return;
    // Once acquired, stay near the known HUD for the rest of this video.
    // Wider local retries back off, while the narrow check above runs per frame.
    if (this.lastAnchor) {
      const retrySeconds = 2 ** Math.min(this.missedSearches, 2);
      this.nextSearch = index + Math.max(1, Math.round(this.fps * retrySeconds));
      this.missedSearches++;
      const hit = this.refine(
        frame,
        this.lastAnchor,
        Math.max(24, 192 * this.lastAnchor.scale),
      );
      if (hit && hit.score >= ICON_MIN_SCORE) {
        this.localRecoveries++;
        this.nextSearch = index + Math.max(1, Math.round(this.fps));
        yield hit;
        return;
      }
      return;
    }
    // Global search is only used until the first successful acquisition.
    this.nextSearch = index + Math.max(1, Math.round(this.fps));
    this.searches++;
    const ratio = Math.min(1, 960 / Math.max(frame.width, frame.height));
    const gray = grayscale(
      resize(
        frame,
        Math.round(frame.width * ratio),
        Math.round(frame.height * ratio),
      ),
    );
    const scales = [
      ...new Set([
        ...Array.from({ length: 29 }, (_, i) => 0.25 * 8 ** (i / 28)),
        1 / 3,
        0.5,
        2 / 3,
        1,
        1.5,
        2,
      ]),
    ].sort((a, b) => a - b);
    const candidates = search(
        gray,
        scales.map((s) => s * ratio),
        undefined,
        2,
      ),
      distinct: Anchor[] = [];
    for (const c of candidates) {
      if (
        !distinct.some(
          (h) => Math.abs(c.x - h.x) + Math.abs(c.y - h.y) < 24 * h.scale,
        )
      )
        distinct.push(c);
      if (distinct.length === 6) break;
    }
    const refined: Anchor[] = [];
    for (const c of distinct) {
      if (c.score < 0.4) break;
      const hit = this.refine(
        frame,
        { ...c, x: c.x / ratio, y: c.y / ratio, scale: c.scale / ratio },
        Math.max(6, 4 / ratio),
      );
      if (hit && hit.score >= ICON_MIN_SCORE) refined.push(hit);
    }
    refined.sort((a, b) => b.score - a.score);
    // Acquiring an icon permanently locks the search area for this video.
    // Require a digit candidate before committing: scenery/UI can resemble an
    // icon while its adjacent band is empty. Try every refined candidate so an
    // unreadable higher-scoring lookalike cannot hide the real counter.
    this.nextSearch = index + Math.max(1, Math.round(this.fps));
    yield* refined;
  }
}
function otsu(values: Float64Array): number {
  const histogram = new Float64Array(256);
  for (const value of values)
    histogram[Math.max(0, Math.min(255, Math.floor(value)))]++;
  let mean = 0;
  for (let i = 0; i < 256; i++) mean += (i * histogram[i]) / values.length;
  let w = 0,
    mu = 0,
    best = -1,
    threshold = 0;
  for (let i = 0; i < 256; i++) {
    const p = histogram[i] / values.length;
    w += p;
    mu += p * i;
    const score = (mean * w - mu) ** 2 / Math.max(w * (1 - w), 1e-10);
    if (score > best) {
      best = score;
      threshold = i;
    }
  }
  return threshold;
}
function binary(image: Image, method: Method): Uint8Array {
  const n = image.width * image.height,
    low = new Float64Array(n),
    gray = new Float64Array(n),
    chroma = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    const r = image.data[i * 3],
      g = image.data[i * 3 + 1],
      b = image.data[i * 3 + 2];
    low[i] = Math.min(r, g, b);
    gray[i] = (r + g + b) / 3;
    chroma[i] = Math.max(r, g, b) - low[i];
  }
  const threshold =
    method === "white_180"
      ? 180
      : method === "gray_otsu"
        ? otsu(gray)
        : Math.max(140, otsu(low));
  return Uint8Array.from({ length: n }, (_, i) =>
    Number(
      method === "gray_otsu"
        ? gray[i] > threshold
        : low[i] > threshold && chroma[i] < 75,
    ),
  );
}
function normalize(
  mask: Uint8Array,
  width: number,
  height: number,
  left: number,
  right: number,
): Uint8Array | null {
  let top = -1,
    bottom = -1;
  for (let y = 0; y < height; y++) {
    let count = 0;
    for (let x = left; x <= right; x++) count += mask[y * width + x];
    if (count >= 4) {
      if (top < 0) top = y;
      bottom = y;
    }
  }
  if (top < 0) return null;
  let x0 = right,
    x1 = left,
    y0 = bottom,
    y1 = top,
    count = 0;
  for (let y = top; y <= bottom; y++)
    for (let x = left; x <= right; x++)
      if (mask[y * width + x]) {
        count++;
        x0 = Math.min(x0, x);
        x1 = Math.max(x1, x);
        y0 = Math.min(y0, y);
        y1 = Math.max(y1, y);
      }
  if (count < 8) return null;
  const glyph = new Uint8Array(600);
  for (let y = 0; y < 30; y++)
    for (let x = 0; x < 20; x++)
      glyph[y * 20 + x] =
        mask[
          (y0 + Math.floor((y * (y1 - y0)) / 29)) * width +
            x0 +
            Math.floor((x * (x1 - x0)) / 19)
        ];
  return glyph;
}
export function readLocatedCounter(
  frame: HudFrame,
  anchor: Anchor,
): number | null {
  const ys = Array.from(
    { length: 31 },
    (_, i) => anchor.y + (i + 4.5) * anchor.scale - 0.5 - (frame.origin?.y ?? 0),
  );
  const xs = Array.from(
    { length: 120 },
    (_, i) => anchor.x + (i + 64.5) * anchor.scale - 0.5 - (frame.origin?.x ?? 0),
  ).filter((x) => x <= frame.width - 1);
  if (ys[0] < 0 || ys.at(-1)! > frame.height - 1 || xs.length < 6 || xs[0] < 0)
    return null;
  const sampled = sample(frame, xs, ys),
    band = {
      ...sampled,
      data: Uint8Array.from(sampled.data, (v) => Math.max(0, Math.min(255, v))),
    };
  const votes = new Set<number>();
  for (const method of methods) {
    const mask = binary(band, method),
      runs: number[][] = [];
    for (let x = 0; x < band.width; x++) {
      let count = 0;
      for (let y = 0; y < band.height; y++) count += mask[y * band.width + x];
      if (count <= 2) continue;
      const last = runs.at(-1);
      if (last && last[1] === x - 1) last[1] = x;
      else runs.push([x, x]);
    }
    const digits: number[] = [];
    let previousEnd = 0,
      valid = true;
    for (const [start, end] of runs) {
      if (digits.length && start - previousEnd > 12) break;
      if (
        (!digits.length && start > 12) ||
        end === band.width - 1 ||
        end - start + 1 < 6 ||
        end - start + 1 > 21
      ) {
        valid = false;
        break;
      }
      const g = normalize(mask, band.width, band.height, start, end);
      if (!g) {
        valid = false;
        break;
      }
      const scores = banks[method]
        .map((t, digit) => {
          let intersection = 0,
            union = 0;
          for (let i = 0; i < 600; i++) {
            intersection += g[i] & t[i];
            union += g[i] | t[i];
          }
          return { digit, score: intersection / Math.max(1, union) };
        })
        .sort((a, b) => b.score - a.score || b.digit - a.digit);
      // A segmented glyph counts as a match even when its best template is weak
      // or ambiguous. The candidate is for review, not a verified damage total.
      digits.push(scores[0].digit);
      previousEnd = end;
    }
    if (valid && digits.length >= 1 && digits.length <= 5)
      votes.add(Number(digits.join("")));
  }
  // Different binarizations may disagree; retain the largest complete reading.
  return votes.size ? Math.max(...votes) : null;
}
export class CounterTracker {
  constructor(private breakOnDecrease = false) {}
  private baseline: number | null = null;
  private pending: number | null = null;
  private repeats = 0;
  private pendingTime = 0;
  update(time: number, value: number | null): DamageEvent | null {
    if (value === null) {
      this.baseline = null;
      this.pending = null;
      this.repeats = 0;
      return null;
    }
    if (this.breakOnDecrease && ((this.baseline !== null && value < this.baseline) ||
        (this.pending !== null && value < this.pending)))
      this.baseline = null;
    if (this.pending === value) this.repeats++;
    else {
      this.pending = value;
      this.pendingTime = time;
      this.repeats = 1;
    }
    if (this.repeats !== 2) return null;
    const previous = this.baseline;
    this.baseline = value;
    return previous !== null && value > previous
      ? {
          time: round(this.pendingTime),
          previous,
          value,
          increase: value - previous,
        }
      : null;
  }
}
export function damageClips(
  events: DamageEvent[],
  duration: number,
  options: DamageOptions & { sampling_fps?: number },
): Clip[] {
  const groups: DamageEvent[][] = [];
  for (const e of events) {
    const g = groups.at(-1);
    if (g && e.time - g.at(-1)!.time <= options.gap_s + 1e-8) g.push(e);
    else groups.push([e]);
  }
  const clips: Clip[] = [];
  for (const g of groups) {
    const a = g[0].time,
      b = g.at(-1)!.time;
    const clip: Clip = {
      start: Math.max(0, a - options.before_s),
      end: Math.min(
        duration,
        Math.max(b + options.after_s, a + 1 / (options.sampling_fps ?? 30)),
      ),
      raw_start: a,
      raw_end: b,
      event_count: g.length,
      peak_dbfs: 0,
    };
    if (clip.end <= clip.start) continue;
    const previous = clips.at(-1);
    if (previous && clip.start <= previous.end + 1e-8) {
      previous.end = Math.max(previous.end, clip.end);
      previous.raw_end = b;
      previous.event_count += g.length;
    } else clips.push(clip);
  }
  return clips;
}

File: go/highlight/damage.go
package highlight

import (
	"context"
	_ "embed"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"sort"
)

//go:embed templates.json
var templateJSON []byte

var methods = []string{"white_180", "gray_otsu", "white_otsu"}

type glyph [600]bool
type icon struct {
	values []float64
	norm   float64
}

var banks, icons = loadTemplates()

func loadTemplates() (map[string][10]glyph, []icon) {
	var data struct {
		Digits map[string]string `json:"digits"`
		Icons  []string          `json:"icons"`
	}
	if err := json.Unmarshal(templateJSON, &data); err != nil {
		panic(err)
	}
	banks := map[string][10]glyph{}
	for _, m := range methods {
		var bank [10]glyph
		for n := range bank {
			b, e := hex.DecodeString(data.Digits[fmt.Sprintf("%s_%d", m, n)])
			if e != nil || len(b) != 75 {
				panic("invalid embedded digit template")
			}
			for i := range bank[n] {
				bank[n][i] = (b[i/8] & (1 << uint(7-i%8))) != 0
			}
		}
		banks[m] = bank
	}
	icons := []icon{}
	for _, s := range data.Icons {
		b, e := hex.DecodeString(s)
		if e != nil || len(b) != 19*32 {
			panic("invalid embedded icon template")
		}
		mean := 0.
		for _, v := range b {
			mean += float64(v)
		}
		mean /= float64(len(b))
		t := icon{values: make([]float64, len(b))}
		for i, v := range b {
			t.values[i] = float64(v) - mean
			t.norm += t.values[i] * t.values[i]
		}
		t.norm = math.Sqrt(t.norm)
		icons = append(icons, t)
	}
	return banks, icons
}

const HUDFrameBytes = 200 * 90 * 3

func iconScore(frame []byte) float64 {
	var strip [19][53]float64
	for y := range strip {
		for x := range strip[y] {
			i := ((20+y*2)*200 + x*2) * 3
			strip[y][x] = (float64(frame[i]) + float64(frame[i+1]) + float64(frame[i+2])) / 3
		}
	}
	best := math.Inf(-1)
	for x := 0; x <= 53-32; x++ {
		mean := 0.
		for y := 0; y < 19; y++ {
			for j := 0; j < 32; j++ {
				mean += strip[y][x+j]
			}
		}
		mean /= 19 * 32
		var centered [608]float64
		norm := 0.
		for y := 0; y < 19; y++ {
			for j := 0; j < 32; j++ {
				v := strip[y][x+j] - mean
				centered[y*32+j] = v
				norm += v * v
			}
		}
		norm = math.Sqrt(norm)
		for _, t := range icons {
			dot := 0.
			for i, v := range centered {
				dot += v * t.values[i]
			}
			best = math.Max(best, dot/math.Max(norm*t.norm, 1e-9))
		}
	}
	return best
}

func otsu(values []float64) int {
	var hist [256]float64
	for _, v := range values {
		hist[uint8(v)]++
	}
	total := float64(len(values))
	muTotal := 0.
	for i, h := range hist {
		muTotal += float64(i) * (h / total)
	}
	w, mu, best := 0., 0., -1.
	threshold := 0
	for i, h := range hist {
		p := h / total
		w += p
		mu += p * float64(i)
		d := muTotal*w - mu
		score := d * d / math.Max(w*(1-w), 1e-10)
		if score > best {
			best = score
			threshold = i
		}
	}
	return threshold
}

func binaryMask(frame []byte, method string) [][]bool {
	low := make([]float64, 31*160)
	gray := make([]float64, len(low))
	chroma := make([]float64, len(low))
	for y := 0; y < 31; y++ {
		for x := 0; x < 160; x++ {
			i := ((24+y)*200 + x) * 3
			r, g, b := frame[i], frame[i+1], frame[i+2]
			j := y*160 + x
			low[j] = float64(min(r, g, b))
			gray[j] = (float64(r) + float64(g) + float64(b)) / 3
			chroma[j] = float64(max(r, g, b)) - low[j]
		}
	}
	threshold := 180.
	if method == "gray_otsu" {
		threshold = float64(otsu(gray))
	}
	if method == "white_otsu" {
		threshold = math.Max(140, float64(otsu(low)))
	}
	mask := make([][]bool, 31)
	for y := range mask {
		mask[y] = make([]bool, 160)
		for x := range mask[y] {
			i := y*160 + x
			if method == "gray_otsu" {
				mask[y][x] = gray[i] > threshold
			} else {
				mask[y][x] = low[i] > threshold && chroma[i] < 75
			}
		}
	}
	return mask
}

func normalize(mask [][]bool, left, right int) (glyph, bool) {
	var out glyph
	first, last := -1, -1
	for y, row := range mask {
		n := 0
		for x := left; x < right; x++ {
			if row[x] {
				n++
			}
		}
		if n >= 4 {
			if first < 0 {
				first = y
			}
			last = y
		}
	}
	if first < 0 {
		return out, false
	}
	minY, maxY, minX, maxX, count := len(mask), -1, right, -1, 0
	for y := first; y <= last; y++ {
		for x := left; x < right; x++ {
			if mask[y][x] {
				minY = min(minY, y)
				maxY = max(maxY, y)
				minX = min(minX, x)
				maxX = max(maxX, x)
				count++
			}
		}
	}
	if count < 8 {
		return out, false
	}
	for y := 0; y < 30; y++ {
		for x := 0; x < 20; x++ {
			sy := minY + int(float64(y)*(float64(maxY-minY)/29))
			sx := minX + int(float64(x)*(float64(maxX-minX)/19))
			if y == 29 {
				sy = maxY
			}
			if x == 19 {
				sx = maxX
			}
			out[y*20+x] = mask[sy][sx]
		}
	}
	return out, true
}

func splitGlyphs(mask [][]bool, left, right int, wide bool) []glyph {
	type run struct{ a, b int }
	runs := []run{}
	for x := left; x < right; x++ {
		n := 0
		for _, row := range mask {
			if row[x] {
				n++
			}
		}
		if n <= 2 {
			continue
		}
		if len(runs) > 0 && runs[len(runs)-1].b == x {
			runs[len(runs)-1].b = x + 1
		} else {
			runs = append(runs, run{x, x + 1})
		}
	}
	if wide {
		for i, j := 0, len(runs)-1; i < j; i, j = i+1, j-1 {
			runs[i], runs[j] = runs[j], runs[i]
		}
	}
	out := []glyph{}
	for _, r := range runs {
		if r.a == left {
			if wide {
				break
			}
			continue
		}
		if r.b-r.a < 6 || r.b-r.a > 21 {
			if wide {
				break
			}
			return nil
		}
		g, ok := normalize(mask, r.a, r.b)
		if !ok {
			return nil
		}
		out = append(out, g)
	}
	if wide {
		for i, j := 0, len(out)-1; i < j; i, j = i+1, j-1 {
			out[i], out[j] = out[j], out[i]
		}
	}
	return out
}

func classify(frame []byte, method string) (int, bool) {
	mask := binaryMask(frame, method)
	gs := splitGlyphs(mask, 0, 160, true)
	if len(gs) < 4 || len(gs) > 5 {
		gs = splitGlyphs(mask, 80, 160, false)
		if len(gs) < 1 || len(gs) > 3 {
			return 0, false
		}
	}
	number := 0
	for _, g := range gs {
		type score struct {
			v float64
			n int
		}
		scores := make([]score, 10)
		for n, t := range banks[method] {
			intersection, union := 0, 0
			for i, v := range g {
				if v && t[i] {
					intersection++
				}
				if v || t[i] {
					union++
				}
			}
			scores[n] = score{float64(intersection) / float64(max(1, union)), n}
		}
		sort.Slice(scores, func(i, j int) bool {
			if scores[i].v == scores[j].v {
				return scores[i].n > scores[j].n
			}
			return scores[i].v > scores[j].v
		})
		if scores[0].v < .68 || scores[0].v-scores[1].v < .05 {
			return 0, false
		}
		number = number*10 + scores[0].n
	}
	return number, true
}

// ReadCounter expects a normalized 200x90 RGB24 HUD crop, not a full frame.
// A nil value is unreadable and must never be substituted with zero.
func ReadCounter(frame []byte) (*int, error) {
	if len(frame) != HUDFrameBytes {
		return nil, fmt.Errorf("expected %d RGB bytes, got %d", HUDFrameBytes, len(frame))
	}
	if iconScore(frame) < .6 {
		return nil, nil
	}
	var value *int
	for _, m := range methods {
		n, ok := classify(frame, m)
		if !ok {
			continue
		}
		if value != nil && *value != n {
			return nil, nil
		}
		v := n
		value = &v
	}
	return value, nil
}

type DamageEvent struct {
	Time     float64 `json:"time"`
	Previous int     `json:"previous"`
	Value    int     `json:"value"`
	Increase int     `json:"increase"`
}
type Reading struct {
	Time  float64 `json:"time"`
	Value *int    `json:"value"`
}

// CounterTracker confirms a reading on its second consecutive observation.
// Unreadable frames and drops reset the baseline without generating an event.
type CounterTracker struct {
	baseline    *int
	pending     *int
	pendingTime float64
	repeats     int
}

func (t *CounterTracker) Update(time float64, value *int) *DamageEvent {
	if value == nil {
		t.baseline = nil
		t.pending = nil
		t.repeats = 0
		return nil
	}
	if t.pending != nil && *t.pending == *value {
		t.repeats++
	} else {
		v := *value
		t.pending = &v
		t.pendingTime = time
		t.repeats = 1
	}
	if t.repeats != 2 {
		return nil
	}
	previous := t.baseline
	v := *value
	t.baseline = &v
	if previous != nil && v > *previous {
		return &DamageEvent{rounded(t.pendingTime, 6), *previous, v, v - *previous}
	}
	return nil
}

func DamageClips(events []DamageEvent, duration float64, o DamageOptions) ([]Clip, error) {
	if err := o.Validate(); err != nil {
		return nil, err
	}
	if err := validateDuration(duration); err != nil {
		return nil, err
	}
	groups := [][]DamageEvent{}
	for i, e := range events {
		if !finite(e.Time) || e.Time < 0 || e.Time > duration || (i > 0 && e.Time < events[i-1].Time) {
			return nil, fmt.Errorf("invalid or unordered damage events")
		}
		if len(groups) > 0 {
			g := groups[len(groups)-1]
			if e.Time-g[len(g)-1].Time <= o.Gap+1e-8 {
				groups[len(groups)-1] = append(g, e)
				continue
			}
		}
		groups = append(groups, []DamageEvent{e})
	}
	clips := []Clip{}
	for _, g := range groups {
		a, b := g[0].Time, g[len(g)-1].Time
		c := Clip{Start: math.Max(0, a-o.Before), End: math.Min(duration, math.Max(b+o.After, a+1/float64(o.FPS))), RawStart: a, RawEnd: b, EventCount: len(g)}
		if c.End <= c.Start {
			continue
		}
		if len(clips) > 0 && c.Start <= clips[len(clips)-1].End+1e-8 {
			p := &clips[len(clips)-1]
			p.End = math.Max(p.End, c.End)
			p.RawEnd = b
			p.EventCount += len(g)
		} else {
			clips = append(clips, c)
		}
	}
	return clips, nil
}

type DamageStats struct {
	SampledFrames    int      `json:"sampled_frames"`
	ReadableFrames   int      `json:"readable_frames"`
	UnreadableFrames int      `json:"unreadable_frames"`
	SamplingFPS      int      `json:"sampling_fps"`
	TimestampBasis   string   `json:"timestamp_basis"`
	Warnings         []string `json:"warnings"`
}
type DamageAnalysis struct {
	Duration float64       `json:"duration_seconds"`
	Events   []DamageEvent `json:"events"`
	Clips    []Clip        `json:"clips"`
	Readings []Reading     `json:"damage_readings"`
	Stats    DamageStats   `json:"damage_stats"`
}

func (t Tools) AnalyzeDamage(ctx context.Context, source string, o DamageOptions) (DamageAnalysis, error) {
	result := DamageAnalysis{Events: []DamageEvent{}, Clips: []Clip{}, Readings: []Reading{}}
	if err := o.Validate(); err != nil {
		return result, err
	}
	media, err := t.Probe(ctx, source)
	if err != nil {
		return result, err
	}
	video, err := media.Video()
	if err != nil {
		return result, err
	}
	result.Duration = media.Duration
	if video.Height <= 0 || math.Abs(float64(video.Width)/float64(video.Height)-16./9) > .02 {
		return result, fmt.Errorf("damage mode requires a full 16:9 image with default HUD")
	}
	filter := fmt.Sprintf("fps=%d:start_time=0,crop=iw*200/3840:ih*90/2160:iw*3320/3840:ih*175/2160,scale=200:90", o.FPS)
	tracker := CounterTracker{}
	count, valid := 0, 0
	err = t.stream(ctx, t.FFmpeg, []string{"-v", "error", "-i", source, "-an", "-vf", filter, "-pix_fmt", "rgb24", "-f", "rawvideo", "-"}, func(r io.Reader) error {
		buf := make([]byte, HUDFrameBytes)
		for {
			_, err := io.ReadFull(r, buf)
			if err == io.EOF {
				break
			}
			if err != nil {
				return fmt.Errorf("incomplete HUD frame: %w", err)
			}
			value, err := ReadCounter(buf)
			if err != nil {
				return err
			}
			time := float64(count) / float64(o.FPS)
			if event := tracker.Update(time, value); event != nil {
				result.Events = append(result.Events, *event)
			}
			if value != nil {
				valid++
			}
			if len(result.Readings) == 0 || !sameValue(result.Readings[len(result.Readings)-1].Value, value) {
				result.Readings = append(result.Readings, Reading{rounded(time, 6), value})
			}
			count++
		}
		return nil
	})
	if err != nil {
		return result, err
	}
	result.Clips, err = DamageClips(result.Events, media.Duration, o)
	result.Stats = DamageStats{count, valid, count - valid, o.FPS, "sampling_grid_from_video_start", []string{"Unreadable HUD may be absent, occluded or unrecognized; it is not treated as zero."}}
	if valid == 0 {
		result.Stats.Warnings = append(result.Stats.Warnings, "No reliable damage readings; check HUD layout and recording clarity.")
	}
	return result, err
}

func sameValue(a, b *int) bool {
	if a == nil || b == nil {
		return a == nil && b == nil
	}
	return *a == *b
}
